<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>JScandy by jameshadfield</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-light.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>

        <img style="" src="img/JScandy.v2.svg" />

        <p></p>

        <h2 class="view"><a href="https://jameshadfield.github.io/JScandy/">Load JScandy</a></h2>
        <h2 class="view"><a href="https://github.com/jameshadfield/JScandy">View the Project on GitHub</a></h2>

<!--
        <ul>
          <li><a href="https://github.com/jameshadfield/JScandy/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/jameshadfield/JScandy/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/jameshadfield/JScandy">View On <strong>GitHub</strong></a></li>
        </ul>
 -->

        <p><strong>Version 0.2.4 &mdash; currently unreleased and in testing.</strong></p>

        <p>Please let us know about bugs or improvements via <a href="https://github.com/jameshadfield/JScandy/issues">issues on the github page</a> or <a href="mailto:jh22@sanger.ac.uk">email james</a>. It would be helpful to include any output from the javascript console in chrome (cmd + alt + j) and steps to reproduce the error.
        </p>



      </header>
      <section>

<h3>Interactive visualization of genome phylogenies</h3>
<p>JScandy is a fully interactive tool to allow visualisation of a phylogenetic tree and associated metadata and genomic information such as recombination blocks, gene-presence/absence or SNP variation. Additionally it allows visualisation of plot files (such as GWAS).</p>
<p>Running is as simple as loading <a href="https://jameshadfield.github.io/JScandy/">this page</a> in Google Chrome and then dragging on your data. That's it.</p>

<h3>Data that you can visualise (and their formats)</h3>
<em>Â¡This section is changing frequently at the moment!</em>
<p/>
<ol>
  <li><strong>Phylogenetic tree</strong> is the backbone of the visualization as it links together all the other data. Trees must be in <a href="http://evolution.genetics.washington.edu/phylip/newicktree.html">Newick format</a> (which is the standard output from software such as <a href="http://sco.h-its.org/exelixis/software.html">RAxML</a>). If you need to convert your tree to a different format try using <a href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a>. Files must end in <em>.tre</em> or <em>.tree</em></li>

  <li><strong>Metadata</strong> can be displayed alongside the tree. A comma seperated value (CSV) file with a header is used (<a href="https://raw.githubusercontent.com/jameshadfield/JScandy/master/example_datasets/gubbins/meta.simple.csv">example</a>). The entries of first column must match the tips of the tree is used and have a header of <em>name</em>, <em>lane</em> or <em>isolate</em>. Values can be strings, numbers or booleans. This file must end in <em>.csv</em></li>

  <li><strong>Genome annotation</strong> in GFF3 format (if you need to convert an embl file to a gff3 file try using <a href="https://www.sanger.ac.uk/resources/software/artemis/">Artemis</a> or install <a href="http://www.ebi.ac.uk/Tools/sfc/emboss_seqret/help/">seqret</a> and convert via <em>seqret -sequence GFF_FILE -feature -fformat embl -fopenfile EMBL_FILE -osformat gff -auto</em>). Currently the second column of entries must be either <em>EMBL</em> or <em>artemis</em>. This file must end in <em>.gff</em></li>

  <li><strong>Genomic data</strong> JScandy is designed to take the output from the following programs,  so any output which uses the same format should (in theory) work fine.

  <ul>
    <li><a href="https://github.com/sanger-pathogens/gubbins">Gubbins</a> (GFF3 format&mdash;each line links a whitespace seperated list of taxa with genomic co-ordinates). If you have old output from gubbins (a _rec.tab file) then there is a <a href="https://github.com/jameshadfield/JScandy/blob/master/scripts/gubbins_tab2gff.py">simple python script</a> which will convert this for you. Currently the second column of each line must be <em>GUBBINS</em>. This file must end in <em>.gff</em></li>
    <li><a href="https://github.com/sanger-pathogens/Roary">Roary</a> output (the gene_presence_absence.csv file). This file must end in <em>.csv</em></li>
    <li>SNP data (VCF)&mdash;<em>coming soon</em></li>
  </ul>
  </li>


  <li><strong>Plot data</strong>
  <ul>
     <li>GWAS results in plink format&mdash;a tab deliminated file with header line similar to<br/>
      #CHR&emsp;SNP&emsp;BP&emsp;minLOG10(P)&emsp;log10(p)&emsp;r^2<br/>
      We use the 3rd column as the genome co-ordinate, the 5th column as the <em>p</em> value and the 6th as the R^2 value. This file must end in <em>.plot</em></li>
    </li>
    <li>TRADIS data&mdash;<em>coming soon</em></li>
  </ul>
  </li>

</ol>


<h3>Example Datasets</h3>
<p>Roary and Gubbins example datasets can be loaded from the JScandy starting page. <em>GWAS example coming soon</em></p>


<h3>Screenshots</h3>
<p><em>Coming soon -- interactive svgs.</em></p>

<h3>Known bugs // 'features'</h3>
<p><ul>
  <li>JScandy only works on <a href="https://www.google.com/chrome/browser/desktop/">Google Chrome!</a></li>
</ul>
</p>


<h3>Install your own (local) copy</h3>
<p>Follow the instructions on the <a href="https://github.com/jameshadfield/JScandy/">Github</a> page to install a local (development) copy. </p>


<h3>Designed and coded by </h3>
<p><a href="https://github.com/jameshadfield/" class="user-mention">james hadfield</a> and <a href="https://github.com/simonrharris/" class="user-mention">Simon Harris</a> and released under a MIT license. JScandy uses a modified version of <a href="https://github.com/PhyloCanvas/PhyloCanvas">PhyloCanvas</a> which is maintained by <a href="https://github.com/richardgoater" class="user-mention">Richard Goater</a> and <a href="https://github.com/jyothishnt" class="user-mention">Jyothish Bhai</a>.
</p>




      </section>
      <footer>




        <p><small>This project is maintained by <a href="https://github.com/jameshadfield">james hadfield</a>. Hosted on GitHub Pages &mdash; Theme modified from <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>

  </body>
</html>
